backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/descriptors/DescriptorUtils.kt

```diff
@@ -75,55 +74,120 @@ internal val IrClass.isArrayWithFixedSizeItems: Boolean
+            BridgeDirection(irClass, BridgeDirectionKind.UNBOX)
 
-internal fun IrFunction.needBridgeTo(target: IrFunction)
-        = (0..this.valueParameters.size + 2).any { needBridgeToAt(target, it) }
+        hasValueTypeAt(index) && overriddenFunction.hasReferenceAt(index) ->
+            // Erase to [Any?].
+            BridgeDirection(null, BridgeDirectionKind.BOX)
 
-internal enum class BridgeDirection {
-    NOT_NEEDED,
-    FROM_VALUE_TYPE,
-    TO_VALUE_TYPE
+        hasReferenceAt(index) && overriddenFunction.hasValueTypeAt(index) ->
+            BridgeDirection(irClass, BridgeDirectionKind.UNBOX)
+
+        else -> BridgeDirection.NONE
+    }
 }
 
-private fun IrFunction.bridgeDirectionToAt(target: IrFunction, index: Int)
-       = when {
-            hasValueTypeAt(index) && target.hasReferenceAt(index) -> BridgeDirection.FROM_VALUE_TYPE
-            hasReferenceAt(index) && target.hasValueTypeAt(index) -> BridgeDirection.TO_VALUE_TYPE
-            else -> BridgeDirection.NOT_NEEDED
+private tailrec fun IrType.erasure(): IrClass =
+        when (val classifier = classifierOrFail) {
+            is IrClassSymbol -> classifier.owner
+            is IrTypeParameterSymbol -> classifier.owner.superTypes.first().erasure()
+            else -> error(classifier)
         }
 
-internal class BridgeDirections(val array: Array<BridgeDirection>) {
-    constructor(parametersCount: Int): this(Array<BridgeDirection>(parametersCount + 3, { BridgeDirection.NOT_NEEDED }))
+internal class BridgeDirections(private val array: Array<BridgeDirection>) {
+    constructor(irFunction: IrSimpleFunction, overriddenFunction: IrSimpleFunction)
+            : this(Array<BridgeDirection>(ParameterIndex.allParametersCount(irFunction)) {
+        irFunction.bridgeDirectionToAt(overriddenFunction, ParameterIndex(it))
+    })
+
+    fun allNotNeeded(): Boolean = array.all { it.kind == BridgeDirectionKind.NONE }
 
-    fun allNotNeeded(): Boolean = array.all { it == BridgeDirection.NOT_NEEDED }
+    private fun getDirectionAt(index: ParameterIndex) = array[index.index]
+
+    val returnDirection get() = getDirectionAt(ParameterIndex.RETURN_INDEX)
+    val dispatchReceiverDirection get() = getDirectionAt(ParameterIndex.DISPATCH_RECEIVER_INDEX)
+    val extensionReceiverDirection get() = getDirectionAt(ParameterIndex.EXTENSION_RECEIVER_INDEX)
+    fun parameterDirectionAt(index: Int) = getDirectionAt(ParameterIndex.map(index))
 
     override fun toString(): String {
         val result = StringBuilder()
         array.forEach {
-            result.append(when (it) {
-                BridgeDirection.FROM_VALUE_TYPE -> 'U' // unbox
-                BridgeDirection.TO_VALUE_TYPE   -> 'B' // box
-                BridgeDirection.NOT_NEEDED      -> 'N' // none
+            result.append(when (it.kind) {
+                BridgeDirectionKind.BOX -> 'B'
+                BridgeDirectionKind.UNBOX -> 'U'
+                BridgeDirectionKind.NONE -> 'N'
```

Comment:
Previous implementation kinda guaranteed that we don't get clashing names for different bridges in the same class. I'm not sure about current one.

Reply:
There is no difference, I just renamed enum values.

Reply:
The only potential clash that I see is for bridges returning `Nothing`, but for those the return type will be different for each bridge, so still no real clash.

Reply:
> I just renamed enum values.

Before: if `bridgeDirection1.toString() == bridgeDirection2.toString()`, then `bridgeDirection1 == bridgeDirection2`.
After: not true anymore.

Reply:
On the other hand, I can't find potential clashes in this version, because binary function signatures include a lot of information.

Reply:
Ah, I got it. There is a difference indeed. Still, the difference is in parameter types, which is included in the signature.